src/charm.py

```diff
@@ -740,6 +721,144 @@ def _gen_wp_config(self):
+        finally start the server. The installation process only run on the leader unit. This
+        operation is idempotence.
+        """
+        self._init_pebble_layer()
+        if self.unit.is_leader():
+            if not self._wp_is_installed():
+                self._wp_install()
+        self._container().start("wordpress")
+
+    @staticmethod
+    def _wp_config_path():
+        return "/var/www/html/wp-config.php"
+
+    def _current_wp_config(self):
+        """Retrieve the current version of wp-config.php from server, return None if not exists"""
+        wp_config_path = self._wp_config_path()
+        container = self._container()
+        if container.exists(wp_config_path):
+            return self._container().pull(wp_config_path).read()
+        return None
+
+    def _push_wp_config(self, wp_config):
+        """Update the content of wp-config.php on server"""
+        self._container().push(self._wp_config_path(), wp_config)
+
+    def _core_reconciliation(self):
+        """Reconciliation process for the WordPress core services, returns True if successful.
+
+        It will fail under the following two circumstances:
+          - Peer relation data not ready
+          - Config doesn't provide valid database information and db relation hasn't
+            been established
+
+        It will check if the current wp-config.php file matches the desired config.
+        If not, update the wp-config.php file.
+
+        It will also check if WordPress is installed (WordPress-related tables exist in db).
+        If not, install WordPress (create WordPress required tables in db).
+
+        If any update is needed, it will stop the apache server first to prevent any requests
+        during the update for security reasons.
+        """
+        if not self._replica_consensus_reached():
+            self.unit.status = WaitingStatus("Waiting for unit consensus")
+            self._stop_server()
+            return False
+        db_config_ready = all(
+            self.model.config[key] for key in
+            ("db_host", "db_name", "db_user", "db_password")
+        )
+        db_relation_ready = all(
+            getattr(self.state, "relation_" + key) for key in
+            ("db_host", "db_name", "db_user", "db_password")
+        )
+        if not db_config_ready and not db_relation_ready:
+            self.unit.status = WaitingStatus("Waiting for db relation")
+            self._stop_server()
+            return False
+        wp_config = self._gen_wp_config()
+        current_wp_config = self._current_wp_config()
+        if wp_config != current_wp_config:
+            self._stop_server()
+            self._push_wp_config(wp_config)
+        self._start_server()
+        return True
+
+    def _reconciliation(self, _event):
+        if not self._container().can_connect():
+            self.unit.status = WaitingStatus("Waiting for pebble")
+            return
+        if not self._core_reconciliation():
```

Comment:
At least for me, it would be much easier to understand this if the `_core_reconcilliation` function would just return whether the core is ready potentially with a reason why it is or not and this function then updates the status and starts or stops the server based on the output rather than mixing setting the status in this function and the `_core_reconiclliation` function

Reply:
each `_*_reconciliation` function is part of the `_reconciliation` that changes the state of application (on/off, configurations) to match the state of charm config / relations. The idea is `_reconciliation` delegates to `_*_reconciliation` and treat them as a blackbox. The `_core_reconciliation` is a little bit different here since it will create some essential configurations that other reconciliation function like `_plugin_reconciliation`, `_theme_reconciliation` needs, so it pass some information (`True`/ `False`) outside. I think I will change this to not passing anything outside and add more checks to other  reconciliation functions. It might cost more computer/io resources but after that all the reconciliation is a complete blackbox.

Reply:
👍